samples/metrics_monitor/src/cttmetrics_i915_pmu.cpp

```diff
@@ -157,34 +157,63 @@ struct metric {
     } end;
 };
 
+#define NUM_SAMPLES_PER_ENGINE 3
+#define PMU_ENGINE_INVALID -1
+
+struct engine_metrics {
+     metric metrics[NUM_SAMPLES_PER_ENGINE]; /* 0 - I915_SAMPLE_BUSY/__I915_PMU_OTHER, 1- I915_SAMPLE_WAIT, 2 - I915_SAMPLE_SEMA */
+     uint64_t num_samples;
+};
+
 struct pmu_metrics {
     int fd;
     int read_format;
     uint64_t num_metrics;
-    struct metric* metrics;
+    uint64_t num_engines;
+    struct engine_metrics* engines_metrics;
 };
 
-static int perf_init(struct pmu_metrics *pm, int num_configs, int* configs)
+static int perf_init(struct pmu_metrics *pm, int num_configs, int configs[][NUM_SAMPLES_PER_ENGINE])
 {
-    int i, res;
+    int i, j, res;
 
     memset(pm, 0, sizeof(struct pmu_metrics));
     pm->fd = -1;
     pm->read_format =
         PERF_FORMAT_TOTAL_TIME_ENABLED |
         PERF_FORMAT_GROUP;
-    pm->metrics = (struct metric*)calloc(num_configs, sizeof(struct metric));
-    if (!pm->metrics)
+    pm->engines_metrics = (struct engine_metrics*)calloc(num_configs, sizeof(struct engine_metrics));
+    if (!pm->engines_metrics)
         return -1;
 
     for (i = 0; i < num_configs; ++i) {
-        if (pm->fd < 0)
-            res = pm->fd = perf_i915_open(configs[i], -1, pm->read_format);
-        else
-            res = perf_i915_open(configs[i], pm->fd, pm->read_format);
-        if (res >= 0) {
-            pm->metrics[pm->num_metrics++].config = configs[i];
+         for (j = 0; j < NUM_SAMPLES_PER_ENGINE; ++j) {
+
+             if (PMU_ENGINE_INVALID == configs[i][j])
+                 continue;
+
+            if (pm->fd < 0)
+                res = pm->fd = perf_i915_open(configs[i][j], -1, pm->read_format);
+            else
+                res = perf_i915_open(configs[i][j], pm->fd, pm->read_format);
+            if (res >= 0) {
+
+                if (I915_PMU_ACTUAL_FREQUENCY == configs[i][j])
```

Comment:
I think we should not differentiate between specific metric here. Actually there are 2 set of metrics:
1. Per engine metrics
2. GPU global metrics
I think we better differentiate between these types.